sycl/doc/extensions/SubGroupBitmaskGroupFunctions/sub_group_bitmask_group_algorithms.asciidoc

```diff
@@ -0,0 +1,253 @@
+ext_oneapi_rank_in_mask would return 5.
+
+
+```c++
+namespace sycl {
+template <int dimensions = 1>
+class nd_item {
+  public:
+...
+  // create a bitmask of the sub-group subset this work-item would be a part of
+  // if the sub-group was split into contiguous groups of size `partition_size`
+  sub_group_mask ext_oneapi_partition_sub_group(size_t partition_size) const;
+
+  // get the bitmask representing the work-items of the sub-group that are
+  // actively executing.
+  sub_group_mask ext_oneapi_active_sub_group_items() const;
+
+  //Returns an integer id representing the calling work-item's position within the provided sub_group_mask.
+  size_t ext_oneapi_rank_in_mask(sub_group_mask mask) const;
+...
+};
+}
+```
+
+possible implementation of `ext_oneapi_rank_in_mask`:
+```c++
+size_t ext_oneapi_rank_in_mask(sub_group_mask mask){
+  // taking 1 from a power of two will give all 1s below the bit of our work-item
+  // anding this with the mask will leave only 1s where there are work-items with lower ids
+  // the popcount of this is the linear id in that sub_group_mask
+  return popcount(mask & ((1 << get_sub_group().get_local_linear_id()) -1));
+}
+```
+
+## Changes to `joint_reduce` and `reduce_over_group`
+Addition of bitmasked version of methods for sub-groups.
+
+```c++
+namespace sycl::ext::oneapi {
+template <typename Ptr, class BinaryOperation>
+joint_reduce(sycl::sub_group, Ptr first, Ptr last, BinaryOperation binary_op, sycl::sub_group_mask);
+
+template <typename Ptr, typename T, class BinaryOperation>
+joint_reduce(sycl::sub_group, Ptr first, Ptr last, T init, BinaryOperation binary_op, sycl::sub_group_mask);
+
+template <typename T, class BinaryOperation>
+reduce_over_group(sycl::sub_group, T x, BinaryOperation binary_op, sycl::sub_group_mask);
+
+template <typename V, typename T, class BinaryOperation>
+reduce_over_group(sycl::sub_group, V x, T init, BinaryOperation binary_op, sycl::sub_group_mask);
+} // namespace sycl::ext::oneapi
+```
+
+## Group_barrier
+New overloads will be required for `sub_group` that takes a bitmask.
+
+```c++
+namespace sycl::ext::oneapi {
+group_barrier(sycl::sub_group, sycl::sub_group_mask mask, memory_scope fence_scope = sub_group::fence_scope);
+} // namespace sycl::ext::oneapi
+```
+
+## New `async_group_copy` function.
+Currently the `nd_item` and `group` classes have the member function
+`async_work_group_copy`. This will be deprecated in favor of a
+`async_group_copy` that will have the group type as a template parameter, much
+like `group_barrier`.
+
+`async_group_copy` methods would be valid for groups `group and `sub_group`.
+For subgroup, a new bitmasked implementation will also be required.
+
+```c++
+namespace sycl::ext::oneapi {
+template <typename Group, typename dataT>
+device_event async_group_copy(Group group, sycl::decorated_local_ptr<dataT> dest, sycl::decorated_global_ptr<dataT> src, size_t numElements);
```

Comment:
How feasible would it be to try and align with `std::future` instead of `sycl::event`?  Would that make sense?

Reply:
As far as the CUDA backend implementation is concerned it doesn't use the list of events at all!  The PTX assembly as it is currently used in our implementation waits on all previous async copy events that were committed to the group that the local thread is part of (the group that is given as an argument to ext_oneapi_wait etc).  Beyond this, an optional argument, N, may be provided to the PTX that allows the group to *not* wait on the previous N event that were committed to the group: i.e. it waits on all committed events other than the last N events.

I think that we decided to use the event interface as it is for consistency with other backends.  But as far as the CUDA backend is concerned it is completely flexible on this question.